Added GUI to Ryujinx

Xpl0itR opened the pull request:
## Initial PR to add a GUI to Ryujinx. Additional features will be available with subsequent PRs.
### Features
- [x] Game list with metadata
  - [x] Title name and ID
  - [x] Developer's name
  - [x] Version number
  - [x] Time played
  - [x] Last played
  - [x] File extension
  - [x] File size
  - [x] File path
- [x] Option in file menu to load a game from a file
- [x] Option in file menu to load a game from a folder
- [x] Option in file menu to open the Ryujinx folder to copy over their keys and stuff
- [x] Option in the options menu to fullscreen the GUI window
- [x] Option in the options menu to open the settings window which contains all options present in the config
- [x] Option in the help menu to open the about window, which contains information about Ryujinx, including links
- [x] Dark Theme by default (no light theme around here)
- [x] Ability to use user defined themes instead of defualt theme
- [x] Dependencies packaged in a nuget package

EmulationChannel commented:
Gui Ryujinx based QT 5.12.0

EmulationChannel commented:
@Xpl0itR 
[Ryujinx.log](https://github.com/Ryujinx/Ryujinx/files/3247073/Ryujinx.log)

EmulationChannel commented:
00:00:00.578 | 0001 Emulation CurrentDomain_UnhandledException: Unhandled exception caught: System.TypeInitializationException: The type initializer for 'Gtk.Application' threw an exception. ---> System.DllNotFoundException: Gtk
   at GLibrary.Load(Library library)
   at Gtk.Application..cctor()
   --- End of inner exception stack trace ---
   at Gtk.Application.Init()
   at Ryujinx.MainMenU.MainMenu(Switch device) in C:\projects\ryujinx\Ryujinx\MainMenu.cs:line 14
   at Ryujinx.Program.Main(String[] args) in C:\projects\ryujinx\Ryujinx\Program.cs:line 58
00:00:00.183 | 0001 Emulation CurrentDomain_UnhandledException: Unhandled exception caught: System.TypeInitializationException: The type initializer for 'Gtk.Application' threw an exception. ---> System.DllNotFoundException: Gtk
   at GLibrary.Load(Library library)
   at Gtk.Application..cctor()
   --- End of inner exception stack trace ---
   at Gtk.Application.Init()
   at Ryujinx.MainMenU.MainMenu(Switch device) in C:\projects\ryujinx\Ryujinx\MainMenu.cs:line 14
   at Ryujinx.Program.Main(String[] args) in C:\projects\ryujinx\Ryujinx\Program.cs:line 58

zephyren25 commented:
@EmulationChannel you need to install GTK

EmulationChannel commented:
@zephyren25 Link download Fron Windows ?

Xpl0itR commented:
@EmulationChannel https://github.com/GtkSharp/GtkSharp/wiki/Installing-Gtk-on-Windows

EmulationChannel commented:
00:00:00.178 | 0001 Emulation CurrentDomain_UnhandledException: Unhandled exception caught: System.TypeInitializationException: The type initializer for 'Gtk.Application' threw an exception. ---> System.DllNotFoundException: Gtk
   at GLibrary.Load(Library library)
   at Gtk.Application..cctor()
   --- End of inner exception stack trace ---
   at Gtk.Application.Init()
   at Ryujinx.MainMenU.MainMenu(Switch device) in C:\projects\ryujinx\Ryujinx\MainMenu.cs:line 14
   at Ryujinx.Program.Main(String[] args) in C:\projects\ryujinx\Ryujinx\Program.cs:line 58
00:00:00.173 | 0001 Emulation CurrentDomain_UnhandledException: Unhandled exception caught: System.TypeInitializationException: The type initializer for 'Gtk.Application' threw an exception. ---> System.DllNotFoundException: Gtk
   at GLibrary.Load(Library library)
   at Gtk.Application..cctor()
   --- End of inner exception stack trace ---
   at Gtk.Application.Init()
   at Ryujinx.MainMenU.MainMenu(Switch device) in C:\projects\ryujinx\Ryujinx\MainMenu.cs:line 14
   at Ryujinx.Program.Main(String[] args) in C:\projects\ryujinx\Ryujinx\Program.cs:line 58

EmulationChannel commented:
Dependencies all installed and updated plus RyujiNX opens

EmulationChannel commented:
00:00:00.173 | 0001 Emulation CurrentDomain_UnhandledException: Unhandled exception caught: System.TypeInitializationException: The type initializer for 'Gtk.Application' threw an exception. ---> System.DllNotFoundException: Gtk
   at GLibrary.Load(Library library)
   at Gtk.Application..cctor()
   --- End of inner exception stack trace ---
   at Gtk.Application.Init()
   at Ryujinx.MainMenU.MainMenu(Switch device) in C:\projects\ryujinx\Ryujinx\MainMenu.cs:line 14
   at Ryujinx.Program.Main(String[] args) in C:\projects\ryujinx\Ryujinx\Program.cs:line 58

Unhandled Exception: System.TypeInitializationException: The type initializer for 'Gtk.Application' threw an exception. ---> System.DllNotFoundException: Gtk
   at GLibrary.Load(Library library)
   at Gtk.Application..cctor()
   --- End of inner exception stack trace ---
   at Gtk.Application.Init()
   at Ryujinx.MainMenU.MainMenu(Switch device) in C:\projects\ryujinx\Ryujinx\MainMenu.cs:line 14
   at Ryujinx.Program.Main(String[] args) in C:\projects\ryujinx\Ryujinx\Program.cs:line 58

emmauss commented:
Did you modify your PATH?

EmulationChannel commented:
@emmauss 
![SharedScreenshot33](https://user-images.githubusercontent.com/37957125/58802894-a3e14d00-85e4-11e9-8ee6-89c396d18322.jpg)

EmulationChannel commented:
[Ryujinx.log](https://github.com/Ryujinx/Ryujinx/files/3247624/Ryujinx.log)

emmauss commented:
change your system PATH, not the user one

EmulationChannel commented:
@emmauss Where to change PATCH system without user

EmulationChannel commented:
@Xpl0itR @zephyren25 @emmauss THANK YOU MY FRIEND
![SharedScreenshot55](https://user-images.githubusercontent.com/37957125/58806826-561d1280-85ed-11e9-93b1-f84177ffea45.jpg)

EmulationChannel commented:
@Xpl0itR New Color Dark
![SharedScreenshotCCC](https://user-images.githubusercontent.com/37957125/58893234-1bd37400-86c6-11e9-922b-b41e19b64bea.jpg)

AcK77 commented:
@EmulationChannel Can you please stop to spam? I think Xpl0itR know the result of what he did! Thanks!

Xpl0itR commented:
@EmulationChannel next time u have an issue and wanna dump logs or just have a question come over to the discord server and ask there so that u don't spam here

Xpl0itR commented:
Well i've added a few new features and cleaned up the code a bit. Main issue is Gtk needing to be installed on your system to use the GUI. No idea how i'm gonna solve this issue rn.

Xpl0itR commented:
Added custom GUI theme support and changed the defualt theme to one I just wrote and also fixed up the todo in the PR's initial comment

Xpl0itR commented:
Added GTK to process path, fixed a bug and minor edits. Required files can be downloaded from [here](https://drive.google.com/open?id=1s-eo3mz7e5HoVy3LrAyu3iSJ4WaKA98C). `bin` and `lib` folders must be in the same directory as `publish` folder

Xpl0itR commented:
Fixed the bug where the GUI crashes due to missing icons, the default icons are now available in the zip. I will replacing the default icons with a more lightweight pack in the future. Added GLArea to main menu and created its binding, also wrote the code that shows/hides the GLArea and the game table when the GUI loads/game starts. Config menu now saves config to memory but still does not save to disk.

Xpl0itR commented:
general settings menu is now fully functional. also fixed the bug where ryujinx crashes when it trys to load an invalid gamedir. Also big rewrite.

Xpl0itR commented:
We got a new Menu! You can now change ur controls from within the GUI. The dictionary in `ControlSettings.cs` is incomplete, will complete it in a future commit.

BaronKiko commented:
So I have made some mock-ups. This is just my opinion so it would be nice to hear from other dev.
Before the mock-ups some general notes:
 - The settings window shouldn't be resizable
 - I have merged the controller setting into the main settings as a tab as it seems more appropriate
 - Games list isn't refreshed when updating sources in settings. You have to reboot
 - Load application through file menu causes crash on my pc
 - Opening settings issues warning about missing icon and theme
 - If you aren't doing meta data for this version then those columns should be removed as they take up a lot of space for no reason atm
 - The About window needs some work, I can do a mock-up of that later
 - The tools menu can be removed as people will just think it's broken
 - Because settings is just 1 option it should probably be moved into file
 - You are storing game dir data in a seperate file when it should be in the config
 - All the options should have tooltips with the description from the json schema
 - You have changed the jsonc file to a json file and remove all the comments. I feel it should still contain the comments when you download it and only lose them when you save within the gui. Then a user can still read what the settings do if they want to edit it manually (it's not common I know but I certainly would still use that and there are really no downsides)

I may have missed something there but that will do for now. Onto the mock-ups.
There are 3 tabs for the settings each with its own mock-up

General:
![general](https://user-images.githubusercontent.com/16206533/59231724-ae4da900-8bd9-11e9-92d8-f56739bb6f5b.png)

Controller, only added a button and drop down here however it should have an extra set of tabs because it's part of the main settings window:
![Photoshop_2019-06-10_23-44-43](https://user-images.githubusercontent.com/16206533/59231775-d0dfc200-8bd9-11e9-89e0-c09517a72eb7.png)

System:
![Photoshop_2019-06-10_23-45-40](https://user-images.githubusercontent.com/16206533/59231798-e8b74600-8bd9-11e9-975b-219d3c442c3a.png)

That's everything for the moment.

gdkchan commented:
I agree with everything said above, except:
```
Because settings is just 1 option it should probably be moved into file
```
I disagree as it doesn't seems to be related to "files" to me (it doesn't open or save anything. well, except the config file, but that can happen at any time, as long its saved before the emulator is closed.

I think some "quick access" options could be added on the `Options` menu aswell, like for example `Full Screen` (possibly with a shortcut also).
```
You have changed the jsonc file to a json file and remove all the comments.
```
I personally think it's fine this way. And imo we should ship the emulator *without* any configuration file. The GUI (or whatever is loading configuration atm) should notice that and load the default configuration, and create a new file.

I still need to actually test the GUI, and will comment later after I do.

BaronKiko commented:
If you can come up with some things to put in the options menu then sure. It just felt pointless with only 1 item.
Fullscreen is a good one, maybe a docked mode toggle too?

If you are going to ship without a config then it should be saved to Appdata so you can share it between builds. However if there is a config in the main ryu folder it should use that config instead as it is useful to have multiple copies of ryu with different settings sometimes.

Another thing I thought of, I'm pretty sure this isn't possible atm because of .net core but in .net core 3 hopefully we can have a locate logs button.
Maybe today we can have a dump logs to documents or something so when you wanna find them it's easier.

gdkchan commented:
So I tested the GUI a bit. Feedback so far:
- When I try to open the "General Settings" menu, I get this exception:
```
Marshaling activate signal
Exception in Gtk# callback delegate
  Note: Applications can use GLib.ExceptionManager.UnhandledException to handle the exception.
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.IO.FileNotFoundException: Could not find file 'C:\Users\Redacted\Documents\Ryujinx\Ryujinx\GameDirs.dat'.
```
- On the controller settings menu, it's possible to press multiple buttons at once. Imo it should be only possible to press one at once (settings multiple buttons to the same key is not useful).
- Still on the controller settings menu, the image on the right looks nice, but there's a bit of aliasing on the borders. Not a big deal, but if possible would be nice to touch it up a bit.
- The about dialog still needs some work as mentioned above. Instead of version, we could add commit hash maybe. Maybe also build date, and name of contributors (or maybe just "Ryujinx Team").

Xpl0itR commented:
In response to Andy's comments about merging the general settings and control setting menus, personally I think that they should be kept separate as nesting a notebook inside a notebook would look ugly imo and doesn't really bring any benefits.
I agree with the idea of having a locate logs button as soon as its possible however i disagree with the temporary method you have suggested. Its just as easy to got to ur ryujinx dir and open the log file as it is going to docs and opening the log file. Personally, as a temp option, i think that the button should open up the file itself in the default text editor

In response to gdk's comments about the image in controller settings, the image has already been replaced with a new one by Ac_k. 

Also the about menu wasn't made using glade but with gtk's default about menu which is very limited which is why it looks bad, i guess it would be a good idea to make one in glade.

Finally, about the error. I accidentally deleted a check when i was rewriting that section, my bad. I will be removing that file anyway in response to Andy's feedback

BaronKiko commented:
> In response to Andy's comments about merging the general settings and control setting menus, personally I think that they should be kept separate as nesting a notebook inside a notebook would look ugly imo and doesn't really bring any benefits.

It does bring the benefit of not having to close the settings just to open another menu which I found annoying. It also consolidates writes. Having used many applications before with multiple settings windows it's really bad design as it can become very hard to find a setting and makes it harder to add features later. For example if you wanted a search field in the setting menu you would have to add it in 2 places.
If your concern is nested notebooks:
1) That's not entirely uncommon and actually standard if your notebook has too many tabs;
2) Remember there should be some spacing with buttons above the controller notebook; and
3) Maybe a notebook for controller number isn't the best choice. You could just have a drop down to select the desired controller for example.
4) Eventually it would be nice to have multiple saved profiles that you can easily swap between and other extra features on the controller window so there will be quite a bit more to add meaning there will be a lot more above the binding part.

> I agree with the idea of having a locate logs button as soon as its possible however i disagree with the temporary method you have suggested. Its just as easy to got to ur ryujinx dir and open the log file as it is going to docs and opening the log file. Personally, as a temp option, i think that the button should open up the file itself in the default text editor
The reason I suggested this is because with an installer the idea is most people wont need to care where there install folder is. Maybe as a simple go between add a text entry for the log file path so they can at least retrieve it. It would be nice to make it editable one day but I would just disable editing for now.

Xpl0itR commented:
1st wave of changes done, more to come soon.
Changes I still need to do:
- Updating games list when game dirs are updated
- Put game dirs into config file
- buttons and new logic for game dir box
- Tooltips
- put more functionality into options menu
- Fix the bug with the controller settings menu
- About Dialog

Xpl0itR commented:
2nd wave of changes done, more to come soon.
Changes I still need to do:
- buttons and new logic for game dir box
- Fix the bug with the controller settings menu
- Better about dialog (waiting on andy's mock-up)

Xpl0itR commented:
3rd wave of changes done, all that's left is the new about dialog

Xpl0itR commented:
merged andy's changes and added new about window. About window still needs styling.

Xpl0itR commented:
Added spin button for new option and tooltips to settings

manotroll commented:
and how to open the graphical interface?

Xpl0itR commented:
@manotroll double click the exe 🤦‍♂

manotroll commented:
> 
> 
> @manotroll double click the exe 🤦‍♂

but I did it in Ryujinx.exe 
there is no other executable in the folder

Xpl0itR commented:
forgot to make this comment earlier, Game icons and names are now shown in the games list

Xpl0itR commented:
Added NuGet package which contains GTK dependencies. You no longer need to download and extract a zip. Should now work out of the box.

Xpl0itR commented:
marked the pr as ready for review

emmauss commented on Ryujinx/GUI/MainMenu.cs:
```diff
@@ -0,0 +1,402 @@
+
+        internal static HLE.Switch device { get; set; }
+
+        private static Application gtkapp { get; set; }
```
properties should not be private. make it a field if it is only used within  class.

emmauss commented on Ryujinx/GUI/MainMenu.cs:
```diff
@@ -0,0 +1,402 @@
+
+        [GUI] Window         MainWin;
+        [GUI] CheckMenuItem  FullScreen;
+        [GUI] MenuItem       NFC;
```
Acronyms should still follow name style. `NFC`  -> `Nfc`. change the id in the glade file too.

emmauss commented on Ryujinx/GUI/MainMenu.cs:
```diff
@@ -0,0 +1,402 @@
+        [GUI] MenuItem       NFC;
+        [GUI] TreeView       GameTable;
+        [GUI] ScrolledWindow GameTableWindow;
+        [GUI] GLArea         GLScreen;
```
`GlScreen`. Also Glarea is not needed right now as embedded rendering still does not work.

emmauss commented on Ryujinx/GUI/MainMenu.cs:
```diff
@@ -0,0 +1,402 @@
+        [GUI] ScrolledWindow GameTableWindow;
+        [GUI] GLArea         GLScreen;
+
+        public MainMenu(string[] args, Application _gtkapp) : this(new Builder("Ryujinx.GUI.MainMenu.glade"), args, _gtkapp) { }
```
camelCase for arguments .

emmauss commented on Ryujinx/GUI/MainMenu.cs:
```diff
@@ -0,0 +1,402 @@
+            settings.XftHinting   = 1;
+            settings.XftHintstyle = "hintfull";
+
+            CssProvider css_provider = new CssProvider();
```
snake_case is not used in C#

emmauss commented on Ryujinx/GUI/MainMenu.cs:
```diff
@@ -0,0 +1,402 @@
+            GameTableWindow.Hide();
+            GLScreen.Show();
+
+            Destroy();
```
Doesn't this close the window. It'd be better to keep the window alive.

Xpl0itR replied:
The window becomes unresponsive after opening the game window so i just destroy it for now until we render to the GLArea

emmauss replied:
That's because you are running it from the main gui, which the gui is running on. Spawn a new thread and call the renderer with that thread.

Xpl0itR replied:
👌

emmauss commented on Ryujinx/GUI/MainMenu.cs:
```diff
@@ -0,0 +1,402 @@
+
+namespace Ryujinx
+{
+    public class MainMenu : Window
```
Change the class name. This is not a menu, but a window

emmauss commented on Ryujinx/GUI/MainMenu.cs:
```diff
@@ -0,0 +1,402 @@
+                GameTableWindow.Hide();
+                GLScreen.Show();
+
+                Destroy();
```
Same

emmauss commented on Ryujinx/GUI/MainMenu.cs:
```diff
@@ -0,0 +1,402 @@
+                GameTableWindow.Hide();
+                GLScreen.Show();
+
+                Destroy();
```
Same

emmauss commented on Ryujinx/GUI/MainMenu.cs:
```diff
@@ -0,0 +1,402 @@
+        {
+            GameTableWindow.Show();
+            GLScreen.Hide();
+            //will also have to write logic to kill running game
```
Should be a TODO, so you don't forget

emmauss commented on Ryujinx/GUI/MainMenu.cs:
```diff
@@ -0,0 +1,402 @@
+
+        private void Profiler_Pressed(object o, EventArgs args)
+        {
+            //Soon™
```
If you are not using the method, then don't attach it to the event

emmauss commented on Ryujinx/GUI/SwitchSettings.cs:
```diff
@@ -0,0 +1,410 @@
+    {
+        internal static Configuration SwitchConfig { get; set; }
+
+        internal HLE.Switch device { get; set; }
```
PascalCase for properties.

emmauss commented on Ryujinx/GUI/SwitchSettings.cs:
```diff
@@ -0,0 +1,410 @@
+
+        internal HLE.Switch device { get; set; }
+
+        private static ListStore GameDirsBoxStore { get; set; }
```
properties are not private, at least to my knowledge

emmauss commented on Ryujinx/GUI/SwitchSettings.cs:
```diff
@@ -0,0 +1,410 @@
+
+        public static void ConfigureSettings(Configuration Instance) { SwitchConfig = Instance; }
+
+        public SwitchSettings(HLE.Switch _device) : this(new Builder("Ryujinx.GUI.SwitchSettings.glade"), _device) { }
```
no `_` in name for arguments/parameters

Xpl0itR commented on Ryujinx/Configuration.cs:
```diff
@@ -193,68 +226,76 @@ public static void Configure(Switch device)
+            Configure(device, Instance);
+        }
+
+        public static void Configure(Switch device, Configuration SwitchConfig)
```
InitialConfigure exists for the things that only get configured when ryujinx starts, such as the logger, and Configure is for everything else

Xpl0itR commented:
requested changes marked resolved have been changed

Xpl0itR commented:
put CreateGameWindow on a new thread and stopped destroying the main window when a game loads

manotroll commented:
I would like to test you do not have the download link anymore

Yvar192 commented:
https://ci.appveyor.com/project/gdkchan/ryujinx/builds/25647242 Builds are up? @manotroll

Xpl0itR commented:
added new features

Xpl0itR commented:
Resolved merge conflicts

manotroll commented:
> 
> 
> https://ci.appveyor.com/project/gdkchan/ryujinx/builds/25647242 Builds are up? @manotroll

I do not know where to download this page and I do not know how to compile

Xpl0itR commented:
Made some changes over the past few days:
- Fixed a lot of bugs
- Rewrote `ApplicationLibrary.cs` (added comments too) so any devs reading it wont get eye cancer
- Added 2 new columns (Developer name and application version) to the game list and wrote the logic for it
- Ryujinx now loads NRO's TitleName and TitleID from the NACP file
- Moved `ApplicationLibrary.cs` from `Ryujinx` to `Ryujinx.HLE` as that is a better place for it
- Moved contents of GUI folder to Ui folder and changed the namespaces of the GUI files from `Ryujinx` to `Ryujinx.Ui`

emmauss commented on Ryujinx.HLE/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,369 @@
+    public class ApplicationLibrary
+    {
+        private static Keyset KeySet;
+        private static HLE.HOS.SystemState.TitleLanguage DesiredTitleLanguage;
```
use a using

emmauss commented on Ryujinx.HLE/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,369 @@
+
+                using (FileStream file = new FileStream(applicationPath, FileMode.Open, FileAccess.Read))
+                {
+                    if ((Path.GetExtension(applicationPath) == ".nsp") || (Path.GetExtension(applicationPath) == ".pfs0") || (Path.GetExtension(applicationPath) == ".xci"))
```
you can split this line, and probably most of the other long lines

emmauss commented on Ryujinx.HLE/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,369 @@
+
+                            // Creates NACP class from the NACP file
+                            IFile controlNacp = controlFs.OpenFile("/control.nacp", OpenMode.Read);
+                            Nacp controlData  = new Nacp(controlNacp.AsStream());
```
align `controlData`

emmauss commented on Ryujinx.HLE/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,369 @@
+
+                        StringBuilder titlename = new StringBuilder();
+                        titlename.Append(Path.GetFileName(applicationPath));
+                        titlename.Remove(Path.GetFileName(applicationPath).Length - Path.GetExtension(applicationPath).Length, Path.GetExtension(applicationPath).Length);
```
line can be split

emmauss commented on Ryujinx/Program.cs:
```diff
@@ -1,151 +1,41 @@
 
-            audioOut.Dispose();
+            Application gtkapp = new Application("Ryujinx.Ryujinx", GLib.ApplicationFlags.None);
+            MainWindow win     = new MainWindow(args, gtkapp);
```
align `win`

emmauss commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,477 @@
+
+            _audioOut = InitializeAudioEngine();
+
+            _device   = new HLE.Switch(_renderer, _audioOut);
```
no need to align if these are not grouped together

emmauss commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,477 @@
+            {
+                //Box.Remove(GameTableWindow);
+
+                //Temporary code section start, remove this section and uncomment above line when game is rendered to the glarea in the gui
```
Remove currently unused code. you can make this a TODO

emmauss commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,477 @@
+                eRrOr.Run();
+                eRrOr.Destroy();
+            }
+
```
unnecessary whitespace

emmauss commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,477 @@
+        {
+            if (_GameLoaded)
+            {
+                MessageDialog eRrOr = new MessageDialog(null, DialogFlags.Modal, MessageType.Error, ButtonsType.Ok, "A game has already been loaded. Please close the emulator and try again");
```
what the heck is this name. please follow naming rules

emmauss commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,477 @@
+            try
+            {
+                string appdataPath = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
+                string savePath = System.IO.Path.Combine(appdataPath, "RyuFs", "nand", "user", "save", "0000000000000000", userId, _device.System.TitleID);
```
align. maybe split the line

emmauss commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,477 @@
+
+            if (fc.Run() == (int)ResponseType.Accept)
+            {
+                //Soon™
```
Make a TODO here

emmauss commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,477 @@
+        {
+            Box.Remove(GlScreen);
+            Box.Add(GameTableWindow);
+            //will also have to write logic to kill running game
```
This can be a TODO

emmauss commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,448 @@
+
+            GameDirsBox.AppendColumn("", new CellRendererText(), "text", 0);
+            _GameDirsBoxStore  = new ListStore(typeof(string));
+            GameDirsBox.Model = _GameDirsBoxStore;
```
align

Xpl0itR commented:
added some new features, updated nuget package with missing dlls and changed requested changes

gdkchan commented on Ryujinx.HLE/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,373 @@
+            DesiredTitleLanguage  = desiredTitleLanguage;
+
+            // Loads the default application Icons
+            using (Stream resourceStream = Assembly.GetExecutingAssembly().GetManifestResourceStream("Ryujinx.HLE.ryujinxNSPIcon.png"))
```
I think this code should be on a separate method, where you pass the resource name, and it returns a byte array. Also imo its better to use the `Stream.Read` method to read the data into a array directly. I'm pretty sure we already have code to do something similar on Ryujinx.HLE (for the profile image for example, used on the account service).

gdkchan commented on Ryujinx.HLE/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,373 @@
+        private static Keyset KeySet;
+        private static HOS.SystemState.TitleLanguage DesiredTitleLanguage;
+
+        public static byte[] RyujinxNspIcon { get; private set; }
```
I think those icons belong to the GUI. They don't have anything to do with the Ryujinx.HLE project.

Xpl0itR replied:
iirc the switch UI does have an application library that shows all the games, with their icons, installed to the sd and the game on an inserted cart. so this is a HLE implementation of that

gdkchan replied:
I guess you're talking about qlaunch. But this doesn't have anything to do with qlaunch. Either way, I don't think what the switch OS does it that important here. From what i've seen, I believe it's better to split content loading into a separate project. Maybe `Ryujinx.Loaders`. Then `Ryujinx.GUI` and `Ryujinx.HLE` would be able to use it to load things. Including getting game icons, loading XCI, "deconstructed" ROMs, etc.

Xpl0itR replied:
btw i was talking about [this](https://upload.wikimedia.org/wikipedia/en/3/34/Nintendo_Switch_Menu_screenshot.png) not qlaunch. But yea i can move it to `Ryujinx.Loaders`

gdkchan replied:
That menu you just linked is called `qlaunch`.

Xpl0itR replied:
oops my bad, i didnt know it was called qlaunch and when i googled it i read a description that didnt seem like the picture

Xpl0itR replied:
well there is no project called `Ryujinx.Loaders`, should I make a new project or put it into `Ryujinx.HLE.Loaders`

gdkchan replied:
I think those icons should be part of the GUI. theres no use for them in any other project.

gdkchan commented on Ryujinx.HLE/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,373 @@
+        public static byte[] RyujinxNroIcon { get; private set; }
+        public static byte[] RyujinxNsoIcon { get; private set; }
+
+        public static List<ApplicationData> ApplicationLibraryData { get; private set; }
```
Same.

Xpl0itR replied:
same as above reply

gdkchan commented on Ryujinx.HLE/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,373 @@
+                            IFileSystem controlFs = null;
+
+                            // Store the ControlFS in variable called controlFs
+                            if (Path.GetExtension(applicationPath) == ".xci")
```
Imo split the code for each extension into a separate method to improve readability.

Xpl0itR replied:
If I split the code for each extension then wouldn't that mean ill have to repeat the same code a few times

gdkchan replied:
How so? you just need to call an extra method, like `LoadXci`, `LoadNsp` etc. But I'm fine keeping it like this for now.

gdkchan commented on Ryujinx.HLE/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,373 @@
+                    {
+                        float timePlayed = float.Parse(sr.ReadLine());
+
+                        if      (timePlayed <= 60.0)    { playedData[0] = $"{timePlayed}s"; }
```
Would be better to use constant for those numbers. Like `60.0`. Could be somethnig like:
```cs
private const double SecondsPerMinute = 60.0;
private const double SecondsPerHour   = SecondsPerMinute * 60;
```
and then:
```cs
if (timePlayed <= SecondsPerMinute) { playedData[0] = $"{timePlayed}s"; }
```
etc. Also I believe should be just `<` rather than `<=` no? If it's exactly 60 seconds, I'd expect it to show "1 minute" (without a s in the end).

gdkchan commented on Ryujinx.HLE/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,373 @@
+                if (File.Exists(Path.Combine(savePath, "LastPlayed.dat")) == false)
+                {
+                    Directory.CreateDirectory(savePath);
+                    using (FileStream file = File.OpenWrite(Path.Combine(savePath, "LastPlayed.dat"))) { file.Write(Encoding.ASCII.GetBytes("Never")); }
```
Too much code on a single line.

gdkchan commented on Ryujinx.HLE/HOS/Horizon.cs:
```diff
@@ -536,18 +534,35 @@ public void LoadProgram(string filePath)
                         if (asetVersion == 0)
```
Hmm, do we have code to parse the homebrew ASET section in 2 places?

Xpl0itR replied:
i don't get what u mean? should we be parsing 2 places for the ASET section or do u mean that we have parsed the ASET section twice in the code?

gdkchan commented on Ryujinx.HLE/HOS/Kernel/Process/ProcessCreationInfo.cs:
```diff
@@ -5,7 +5,7 @@ struct ProcessCreationInfo
 
         public int  Category { get; private set; }
-        public long TitleId  { get; private set; }
+        public ulong TitleId { get; private set; }
```
Alignment.

gdkchan commented on Ryujinx.HLE/Ryujinx.HLE.csproj:
```diff
@@ -28,11 +28,21 @@
 
   <ItemGroup>
     <None Remove="Homebrew.npdm" />
+    <None Remove="ryujinxNCAIcon.png" />
```
Those resources should be part of the GUI.

gdkchan commented on Ryujinx/Program.cs:
```diff
@@ -1,151 +1,41 @@
-            Configuration.Configure(device);
-
-            Profile.Initialize();
+            string systemPATH = Environment.GetEnvironmentVariable("Path", EnvironmentVariableTarget.Machine);
```
Nit: `systemPATH` -> `systemPath`.

gdkchan commented on Ryujinx/Ryujinx.csproj:
```diff
@@ -18,8 +19,23 @@
+    <EmbeddedResource Include="Ui\assets\GitHubLogo.png" />
+    <EmbeddedResource Include="Ui\assets\JoyCon.png" />
+    <EmbeddedResource Include="Ui\assets\PatreonLogo.png" />
+    <EmbeddedResource Include="Ui\assets\ryujinxIcon.png" />
```
Nit: `ryujinxIcon` -> `RyujinxIcon` (rename).

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,465 @@
+
+        private static IAalOutput _audioOut;
+
+        internal static HLE.Switch _device;
```
Why not just `Switch`? Is the name conflicting with some other name?

Xpl0itR replied:
Yea, conflict with gtk

gdkchan replied:
Ok. I think at some point we should get rid of the Switch class and keep Horizon only. Or maybe replace it with something else.

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,465 @@
+
+        private static Application _gtkapp;
+
+        private static ListStore _TableStore;
```
`_tableStore`.

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,465 @@
+
+        private static ListStore _TableStore;
+
+        private static bool _GameLoaded = false;
```
`_gameLoaded`

gdkchan commented on Ryujinx/Ui/AboutWindow.cs:
```diff
@@ -0,0 +1,90 @@
+    public class AboutWindow : Window
+    {
+#pragma warning disable 649
+        [GUI] Window AboutWin;
```
Any reason for the name to not match our current conventions? It's supposed to be camelCase starting with a underscore.

Xpl0itR replied:
That is how emmauss did it in his old PR so I assumed thats how u wanted it done

gdkchan replied:
I think that PR was before we changed the projects to match the C# naming conventions.

emmauss replied:
I was not sure how the gtk builder handles names when creating the object, ie, if a different case was used for the id and the variable, whether it would bind. so, just to play it safe, I used same cases for both of them. at that time, everything was PascalCase, so, it should be changed now.

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,465 @@
+                        LargeImageText = "Ryujinx is an emulator for the Nintendo Switch"
+                    },
+                    Details    = "Main Menu",
+                    State      = "Idling",
```
Idle.

Xpl0itR replied:
Idling is the present progressive form of the word idle, personally i think it fits better here

gdkchan replied:
Well I'm not a native english speaker, so I'm not going to argue about that. But I usually see `Idle` not `Idling` on other applications. `Idling` sounds weird to me.

Xpl0itR replied:
tbh i was thinking about replacing that with the ryujinx version when i write the logic for that in the next pr

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,465 @@
+            }
+            else
+            {
+                cssProvider.LoadFromPath(System.IO.Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Theme.css"));
```
Add `using System.IO` and then `System.IO.Path` -> `Path`.

Xpl0itR replied:
there is already a using for System.IO however there is a conflict with Gtk over the Path class

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,465 @@
+        {
+            if (_GameLoaded)
+            {
+                MessageDialog errorDialog = new MessageDialog(null, DialogFlags.Modal, MessageType.Error, ButtonsType.Ok, "A game has already been loaded. Please close the emulator and try again");
```
You can't open a new game when theres one currently running? That's weird. I'd expect it to close the current game and open the new one.

gdkchan replied:
Sounds like it would be nice to write a helper method for error messages. Similar to winforms `MessageBox.Show`.

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,465 @@
+                }
+                else
+                {
+                    Logger.PrintWarning(LogClass.Application, "Please specify a valid XCI/NCA/NSP/PFS0/NRO file");
```
This should show a warning message, rather than a entry on the log.

Xpl0itR replied:
The GUI will not show any files apart from XCI/NCA/NSP/PFS0/NRO/NSO files. this error exists for people using ryujinx from a terminal

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,465 @@
+                try
+                {
+                    string appdataPath = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
+                    string savePath = System.IO.Path.Combine(appdataPath, "RyuFs", "nand", "user", "save", "0000000000000000", userId, _device.System.TitleID);
```
Path names shouldn't be hardcoded like this. IIRC we had some string constants with those names. Otherwise if they are changed in one place, then the GUI will break.

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,465 @@
+                string userId = "00000000000000000000000000000001";
+                try
+                {
+                    string appdataPath = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
```
We had some helper methods to build those paths. I don't think the same code should be replicated here. It's better to centralise it all on a single place to avoid divergence and bugs.

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,465 @@
+                    string appdataPath = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
+                    string savePath = System.IO.Path.Combine(appdataPath, "RyuFs", "nand", "user", "save", "0000000000000000", userId, _device.System.TitleID);
+
+                    if (File.Exists(System.IO.Path.Combine(savePath, "TimePlayed.dat")) == false)
```
`System.IO.Path` -> `Path`.

Xpl0itR replied:
There is a conflict with Gtk over the Path class

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,465 @@
+            _audioOut.Dispose();
+            DiscordClient.Dispose();
+            Logger.Shutdown();
+            Environment.Exit(0);
```
Why is that necessary? It can't exit cleanly?

Xpl0itR replied:
If this function is called from the main thread before a game is loaded then it does nothing until a game is loaded and then closes. If this function is called from the main thread after a game has loaded then it works fine. If this function is called from the game window thread then only the game window closes, closing the main window after this causes it to crash.

gdkchan commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,465 @@
+        }
+        
+        //Events
+        private void Row_Activated(object o, RowActivatedArgs args)
```
Why the underscores on the method names?

Xpl0itR replied:
Because they are methods bound to events. I assumed this is the correct way since that how it has been done for the `CurrentDomain_ProcessExit` and `CurrentDomain_UnhandledException` methods in `Program.cs`

gdkchan commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,448 @@
+            ZR1.Label           = SwitchConfig.KeyboardControls.RightJoycon.ButtonZr.ToString();
+
+            CustThemeDir.Buffer.Text  = SwitchConfig.CustomThemePath;
+            FGALMSpinAdjustment.Value = SwitchConfig.FsGlobalAccessLogMode;
```
`FGALM` sounds weird. I'd just call it `FsLog`.

gdkchan commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,448 @@
+                _GameDirsBoxStore.AppendValues(GameDir);
+            }
+
+            if (CustThemeToggle.Active == false) { CustThemeDir.Sensitive = false; CustThemeDirLabel.Sensitive = false; BrowseThemeDir.Sensitive = false; }
```
Too much code on a single line.

gdkchan commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,448 @@
+            GameDirsBox.AppendColumn("", new CellRendererText(), "text", 0);
+            _GameDirsBoxStore  = new ListStore(typeof(string));
+            GameDirsBox.Model  = _GameDirsBoxStore;
+            foreach (string GameDir in SwitchConfig.GameDirs)
```
Not following naming conventions. Most of the code here isn't. `GameDir` -> `gameDir`.

gdkchan commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,448 @@
+                {
+                    string key = KeyPressed.Event.Key.ToString();
+
+                    if (Enum.IsDefined(typeof(OpenTK.Input.Key), key.First().ToString().ToUpper() + key.Substring(1))) { Button.Label = key.First().ToString().ToUpper() + key.Substring(1); }
```
Too much code on a single line, plus repeated code. `key.First().ToString().ToUpper() + key.Substring(1)`.

gdkchan replied:
Imo you should break that down into multiple lines. Cramming everything on a single line like this just makes the code harder to read.

gdkchan commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,448 @@
+
+        private void CustThemeToggle_Activated(object obj, EventArgs args)
+        {
+            if (CustThemeToggle.Active == false) { CustThemeDir.Sensitive = false; CustThemeDirLabel.Sensitive = false; BrowseThemeDir.Sensitive = false; } else { CustThemeDir.Sensitive = true; CustThemeDirLabel.Sensitive = true; BrowseThemeDir.Sensitive = true; }
```
Too much code on a single line.

gdkchan commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,448 @@
+            if (DirectKeyboardAccess.Active)          { SwitchConfig.EnableKeyboard            = true;  }
+            if (CustThemeToggle.Active)               { SwitchConfig.EnableCustomTheme         = true;  }
+
+            if (IconToggle.Active           == false) { SwitchConfig.GuiColumns[0]             = false; }
```
`if (IconToggle.Active           == false)` -> `if (!IconToggle.Active)`. Same for the others below. Could also remove the curly braces I guess.

gdkchan commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,448 @@
+            Destroy();
+        }
+
+        public readonly Dictionary<string, string> GdkToTKInput = new Dictionary<string, string>()
```
`GdkToTKInput` -> `GdkToOpenTKInput`. Doing this kind of key name conversion looks fragile. I still need to look where this is used.

Xpl0itR commented:
Requested changes have been changed

ReaperOfSouls1909 commented:
you asked me to remind you when you where near finsihed about adding the ability to have fav games

Xpl0itR commented:
Ahh yes, I forgot about that. I'll add it to the feature list for the next PR

Xpl0itR commented:
minor updates:
- added update check button that runs ryujinx-updater if the updater is found
- reads version info from installer if found and displays it in about menu

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                                IDirectory controlDir = controlFs.OpenDirectory("./", OpenDirectoryMode.All);
+                                foreach (DirectoryEntry entry in controlDir.Read())
+                                {
+                                    if (entry.Name == "control.nacp") { continue; }
```
Expand brackets here.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                                        applicationIcon = ms.ToArray();
+                                    }
+
+                                    if (applicationIcon != null) { break; }
```
Expand brackets here.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+
+                                if (applicationIcon == null)
+                                {
+                                    if (Path.GetExtension(applicationPath) == ".xci") { applicationIcon = RyujinxXciIcon; }
```
Expand brackets here and in the `else` too.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                                    if (entry.Name == "control.nacp") { continue; }
+
+                                    IFile icon = controlFs.OpenFile(entry.FullPath, OpenMode.Read);
+                                    using (MemoryStream ms = new MemoryStream())
```
```suggestion
                                    using (MemoryStream stream = new MemoryStream())
```
use an explicit name is maybe better.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                            developer = "Unknown";
+                            version   = "?";
+
+                            if (Path.GetExtension(applicationPath) == ".xci") { applicationIcon = RyujinxXciIcon; }
```
Expand brackets here and in the else too.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                            developer = "Unknown";
+                            version   = "?";
+
+                            if (Path.GetExtension(applicationPath) == ".xci") { applicationIcon = RyujinxXciIcon; }
```
Expand brackets here and in the else too.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                            Logger.PrintError(LogClass.Application, $"The file is not an NCA file or the header key is incorrect. Errored File: {applicationPath}");
+                        }
+                    }
+
```
Nit: remove the extra empty line

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                        byte[] Read(long Position, int Size)
+                        {
+                            file.Seek(Position, SeekOrigin.Begin);
+                            return reader.ReadBytes(Size);
```
Add a new empty line before the `return`

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                                }
+
+                                titleId = controlData.PresenceGroupId.ToString("x16");
+                                if (string.IsNullOrWhiteSpace(titleId)) { titleId = controlData.SaveDataOwnerId.ToString("x16"); }
```
Expand brackets here and in the next `if` too.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                                }
+                            }
+                        }
+                        else { applicationIcon = RyujinxNroIcon; titleName = "Application"; titleId = "0000000000000000"; developer = "Unknown"; version = "?"; }
```
Expand brackets here.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                        }
+                        else { applicationIcon = RyujinxNroIcon; titleName = "Application"; titleId = "0000000000000000"; developer = "Unknown"; version = "?"; }
+                    }
+
```
remove extra empty line here

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                    // If its an NCA or NSO we just set defaults
+                    else if ((Path.GetExtension(applicationPath) == ".nca") || (Path.GetExtension(applicationPath) == ".nso"))
+                    {
+                             if (Path.GetExtension(applicationPath) == ".nca") { applicationIcon = RyujinxNcaIcon; }
```
Expand brackets here and remove the `if` alignment.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                    else if ((Path.GetExtension(applicationPath) == ".nca") || (Path.GetExtension(applicationPath) == ".nso"))
+                    {
+                             if (Path.GetExtension(applicationPath) == ".nca") { applicationIcon = RyujinxNcaIcon; }
+                        else if (Path.GetExtension(applicationPath) == ".nso") { applicationIcon = RyujinxNsoIcon; }
```
Expand brackets here.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+
+                        string fileName = Path.GetFileName(applicationPath);
+                        string fileExt  = Path.GetExtension(applicationPath);
+                        StringBuilder titlename = new StringBuilder();
```
Add a new line before.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                if (File.Exists(Path.Combine(savePath, "TimePlayed.dat")) == false)
+                {
+                    Directory.CreateDirectory(savePath);
+                    using (FileStream file = File.OpenWrite(Path.Combine(savePath, "TimePlayed.dat"))) { file.Write(Encoding.ASCII.GetBytes("0")); }
```
Expand brackets here.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                    {
+                        float timePlayed = float.Parse(sr.ReadLine());
+
+                             if (timePlayed < SecondsPerMinute) { playedData[0] = $"{timePlayed}s"; }
```
Expand all condition brackets here and remove the `if` aligments.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+                        file.Write(Encoding.ASCII.GetBytes("Never"));
+                    }
+                }
+                using (FileStream fs = File.OpenRead(Path.Combine(savePath, "LastPlayed.dat")))
```
Nit: Add an empty line before.

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,374 @@
+
+                return playedData;
+            }
+            catch { return new string[] { "Unknown", "Unknown" }; }
```
Expand brackets here.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+
+        private static IAalOutput _audioOut;
+
+        internal static HLE.Switch _device;
```
at least put it at the beginning of vars since it's `internal` and not `private` and then put public vars after.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+        private static bool _gameLoaded = false;
+
+#pragma warning disable 649
+        [GUI] Window         _mainWin;
```
```suggestion
        [GUI] Window        _mainWin;
```
There is an extra empty space in all those `[GUI]` lines.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+
+        internal static HLE.Switch _device;
+
+        private static Application _gtkapp;
```
```suggestion
        private static Application _gtkApplication;
```
Is maybe better to explicit the name.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+
+        public MainWindow(string[] args, Application gtkapp) : this(new Builder("Ryujinx.Ui.MainWindow.glade"), args, gtkapp) { }
+
+        private MainWindow(Builder builder, string[] args, Application gtkapp) : base(builder.GetObject("_mainWin").Handle)
```
same about `gtkapp`

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+            if (args.Length == 1)
+            {
+                // Temporary code section start, remove this section when game is rendered to the GLArea in the GUI
+                _box.Remove(_glScreen);
```
Add an new empty line after.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+                if (SwitchSettings.SwitchConfig.GuiColumns[6]) { _gameTable.AppendColumn("File Ext"   , new CellRendererText()  , "text"  , 6); }
+                if (SwitchSettings.SwitchConfig.GuiColumns[7]) { _gameTable.AppendColumn("File Size"  , new CellRendererText()  , "text"  , 7); }
+                if (SwitchSettings.SwitchConfig.GuiColumns[8]) { _gameTable.AppendColumn("Path"       , new CellRendererText()  , "text"  , 8); }
+                _tableStore      = new ListStore(typeof(Gdk.Pixbuf), typeof(string), typeof(string), typeof(string), typeof(string), typeof(string), typeof(string), typeof(string), typeof(string));
```
Add a new empty line before.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+                if (SwitchSettings.SwitchConfig.GuiColumns[8]) { _gameTable.AppendColumn("Path"       , new CellRendererText()  , "text"  , 8); }
+                _tableStore      = new ListStore(typeof(Gdk.Pixbuf), typeof(string), typeof(string), typeof(string), typeof(string), typeof(string), typeof(string), typeof(string), typeof(string));
+                _gameTable.Model = _tableStore;
+                UpdateGameTable();
```
Add a new empty line before.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+            {
+                // Temporary code section start, remove this section when game is rendered to the GLArea in the GUI
+                _box.Remove(_glScreen);
+                if (SwitchSettings.SwitchConfig.GuiColumns[0]) { _gameTable.AppendColumn("Icon"       , new CellRendererPixbuf(), "pixbuf", 0); }
```
```suggestion
                if (SwitchSettings.SwitchConfig.GuiColumns[0]) { _gameTable.AppendColumn("Icon",        new CellRendererPixbuf(), "pixbuf", 0); }
```
The comma need to be at the end of the args. The alignment is after the comma.
Ditto on the next lines.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+            {
+                _box.Remove(_glScreen);
+
+                if (SwitchSettings.SwitchConfig.GuiColumns[0]) { _gameTable.AppendColumn("Icon"       , new CellRendererPixbuf(), "pixbuf", 0); }
```
```suggestion
                if (SwitchSettings.SwitchConfig.GuiColumns[0]) { _gameTable.AppendColumn("Icon",        new CellRendererPixbuf(), "pixbuf", 0); }
```

The comma need to be at the end of the args. The alignment is after the comma.
Ditto on the next lines.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+                            break;
+                        default:
+                            Logger.PrintInfo(LogClass.Application, "Loading as homebrew.");
+                            try { _device.LoadProgram(path); }
```
Expand brackets here and in the `catch`

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+                    if (File.ReadAllLines(System.IO.Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "RPsupported.dat")).Contains(_device.System.TitleID))
+                    {
+                        DiscordPresence.Assets.LargeImageKey = _device.System.TitleID;
+                    }
```
Add new empty line after.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+                    DiscordClient.SetPresence(DiscordPresence);
+                }
+
+                string userId = "00000000000000000000000000000001";
```
Add new empty line after.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+                    if (File.Exists(System.IO.Path.Combine(savePath, "TimePlayed.dat")) == false)
+                    {
+                        Directory.CreateDirectory(savePath);
+                        using (FileStream file = File.OpenWrite(System.IO.Path.Combine(savePath, "TimePlayed.dat"))) { file.Write(Encoding.ASCII.GetBytes("0")); }
```
Expand brackets her and rename `file` to `stream`.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+                    if (File.Exists(System.IO.Path.Combine(savePath, "LastPlayed.dat")) == false)
+                    {
+                        Directory.CreateDirectory(savePath);
+                        using (FileStream file = File.OpenWrite(System.IO.Path.Combine(savePath, "LastPlayed.dat"))) { file.Write(Encoding.ASCII.GetBytes("Never")); }
```
Ditto

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+                        Directory.CreateDirectory(savePath);
+                        using (FileStream file = File.OpenWrite(System.IO.Path.Combine(savePath, "LastPlayed.dat"))) { file.Write(Encoding.ASCII.GetBytes("Never")); }
+                    }
+                    using (FileStream fs = File.OpenWrite(System.IO.Path.Combine(savePath, "LastPlayed.dat")))
```
Add a new line before and rename `fs` to `stream`

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+
+        private static void End()
+        {
+            string userId = "00000000000000000000000000000001";
```
Maybe you could create a const for that? (Since you use it twice)

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+            {
+                try
+                {
+                    string savePath = System.IO.Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "RyuFS", "nand", "user", "save", "0000000000000000", userId, _device.System.TitleID);
```
Align var assignments

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+                    string savePath = System.IO.Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "RyuFS", "nand", "user", "save", "0000000000000000", userId, _device.System.TitleID);
+                    double currentPlayTime = 0;
+
+                    using (FileStream fs = File.OpenRead(System.IO.Path.Combine(savePath, "LastPlayed.dat")))
```
`fs` > `stream`

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+
+                    using (FileStream fs = File.OpenRead(System.IO.Path.Combine(savePath, "LastPlayed.dat")))
+                    {
+                        using (StreamReader sr = new StreamReader(fs))
```
`sr` > `reader`

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+                        {
+                            DateTime startTime = DateTime.Parse(sr.ReadLine());
+
+                            using (FileStream lpfs = File.OpenRead(System.IO.Path.Combine(savePath, "TimePlayed.dat")))
```
`lpfs` > `l?p?Stream`

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+
+                            using (FileStream lpfs = File.OpenRead(System.IO.Path.Combine(savePath, "TimePlayed.dat")))
+                            {
+                                using (StreamReader lpsr = new StreamReader(lpfs))
```
`lpsr` > `l?p?Reader`

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+                                }
+                            }
+
+                            using (FileStream tpfs = File.OpenWrite(System.IO.Path.Combine(savePath, "TimePlayed.dat")))
```
`tpfs ` > `timePlayedStream`

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+
+                            using (FileStream tpfs = File.OpenWrite(System.IO.Path.Combine(savePath, "TimePlayed.dat")))
+                            {
+                                using (StreamWriter tpsr = new StreamWriter(tpfs))
```
`tpsr` > `timePlayedWriter`

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+        //Events
+        private void Row_Activated(object o, RowActivatedArgs args)
+        {
+            _tableStore.GetIter(out TreeIter treeiter, new TreePath(args.Path.ToString()));
```
```suggestion
            _tableStore.GetIter(out TreeIter treeIter, new TreePath(args.Path.ToString()));
```

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+        private void Load_Application_File(object o, EventArgs args)
+        {
+            FileChooserDialog fc = new FileChooserDialog("Choose the file to open", this, FileChooserAction.Open, "Cancel", ResponseType.Cancel, "Open", ResponseType.Accept);
+            fc.Filter            = new FileFilter();
```
Don't align assign here and add an empty line before.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+        {
+            Process.Start(new ProcessStartInfo()
+            {
+                FileName = System.IO.Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "RyuFs"),
```
Align assignments

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+            });
+        }
+
+        private void Exit_Pressed(object o, EventArgs args) { End(); }
```
Expand brackets.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+
+        private void Exit_Pressed(object o, EventArgs args) { End(); }
+
+        private void Window_Close(object o, DeleteEventArgs args) { End(); }
```
Expand brackets.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+
+        private void FullScreen_Toggled(object o, EventArgs args)
+        {
+            if (_fullScreen.Active == true) { Fullscreen();   }
```
Expand brackets and in the `else` too.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+
+        private void Settings_Pressed(object o, EventArgs args)
+        {
+            SwitchSettings SettingsWin = new SwitchSettings(_device);
```
Add empty line after.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+            SwitchSettings SettingsWin = new SwitchSettings(_device);
+            _gtkapp.Register(GLib.Cancellable.Current);
+            _gtkapp.AddWindow(SettingsWin);
+            SettingsWin.Show();
```
Add empty line before.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+        private void Nfc_Pressed(object o, EventArgs args)
+        {
+            FileChooserDialog fc = new FileChooserDialog("Choose the file to open", this, FileChooserAction.Open, "Cancel", ResponseType.Cancel, "Open", ResponseType.Accept);
+            fc.Filter            = new FileFilter();
```
Don't align assign here and add an empty line before.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+        private void Update_Pressed(object o, EventArgs args)
+        {
+            string ryuUpdater = System.IO.Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "RyuFS", "RyuUpdater.exe");
+            try { Process.Start(new ProcessStartInfo(ryuUpdater, "/U") { UseShellExecute = true }); }
```
Add new line before and expand brackets of the `try / catch`

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+
+        private void About_Pressed(object o, EventArgs args)
+        {
+            AboutWindow AboutWin = new AboutWindow();
```
Add empty line after.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+            AboutWindow AboutWin = new AboutWindow();
+            _gtkapp.Register(GLib.Cancellable.Current);
+            _gtkapp.AddWindow(AboutWin);
+            AboutWin.Show();
```
Add empty line before.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,564 @@
+
+        private void Icon_Toggled(object o, EventArgs args)
+        {
+            if (_iconToggle.Active) SwitchSettings.SwitchConfig.GuiColumns[0] = true;
```
```suggestion
            SwitchSettings.SwitchConfig.GuiColumns[0] = _iconToggle.Active;
```

And same for other `Toggled` method.

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,431 @@
+            _fsLogSpinAdjustment.Value           = SwitchConfig.FsGlobalAccessLogMode;
+
+            _gameDirsBox.AppendColumn("", new CellRendererText(), "text", 0);
+            _gameDirsBoxStore  = new ListStore(typeof(string));
```
```suggestion
            _gameDirsBoxStore   = new ListStore(typeof(string));
```

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,431 @@
+            if (_listeningForKeypress == false)
+            {
+                KeyPressEvent += On_KeyPress;
+                _listeningForKeypress = true;
```
Add new line before

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,431 @@
+                    string key    = KeyPressed.Event.Key.ToString();
+                    string capKey = key.First().ToString().ToUpper() + key.Substring(1);
+
+                    if (Enum.IsDefined(typeof(OpenTK.Input.Key), capKey)) { Button.Label = capKey;                }
```
Expand conditions brackets

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,431 @@
+                    else                                                  { Button.Label = "Space";               }
+
+                    Button.SetStateFlags(0, true);
+                    KeyPressEvent -= On_KeyPress;
```
Add new line before and after

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,431 @@
+                    _listeningForKeypress = false;
+                }
+            }
+            else { Button.SetStateFlags(0, true); }
```
Expand brackets

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,431 @@
+
+        private void AddDir_Pressed(object obj, EventArgs args)
+        {
+            if (Directory.Exists(_addGameDirBox.Buffer.Text)) { _gameDirsBoxStore.AppendValues(_addGameDirBox.Buffer.Text); }
```
Expand brackets

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,431 @@
+
+        private void BrowseDir_Pressed(object obj, EventArgs args)
+        {
+            FileChooserDialog fc = new FileChooserDialog("Choose the game directory to add to the list", this, FileChooserAction.SelectFolder, "Cancel", ResponseType.Cancel, "Add", ResponseType.Accept);
```
`fc` > `fileChooser`

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,431 @@
+        {
+            TreeSelection selection = _gameDirsBox.Selection;
+
+            selection.GetSelected(out TreeIter treeiter);
```
Remove new line before and add a new line after.

```suggestion
            selection.GetSelected(out TreeIter treeIter);
```

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,431 @@
+
+        private void CustThemeToggle_Activated(object obj, EventArgs args)
+        {
+            if (_custThemeToggle.Active == false)
```
```suggestion
            _custThemePath.Sensitive      = _custThemeToggle.Active;
            _custThemePathLabel.Sensitive = _custThemeToggle.Active;
            _browseThemePath.Sensitive    = _custThemeToggle.Active;
```

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,431 @@
+
+        private void BrowseThemeDir_Pressed(object obj, EventArgs args)
+        {
+            FileChooserDialog fc = new FileChooserDialog("Choose the theme to load", this, FileChooserAction.Open, "Cancel", ResponseType.Cancel, "Select", ResponseType.Accept);
```
`fc` > `fileChooser` and add a new line after.

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,431 @@
+        {
+            List<string> gameDirs = new List<string>();
+
+            _gameDirsBoxStore.GetIterFirst(out TreeIter iter);
```
```suggestion
            _gameDirsBoxStore.GetIterFirst(out TreeIter treeIter);
```

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,431 @@
+                _gameDirsBoxStore.IterNext(ref iter);
+            }
+
+            if (_errorLogToggle.Active)        SwitchConfig.LoggingEnableError        = true;
```
```suggestion
            SwitchConfig.LoggingEnableError        = _errorLogToggle.Active;
```
and same for other lines.

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,431 @@
+
+            Configuration.SaveConfig(SwitchConfig, System.IO.Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Config.json"));
+            Configuration.Configure(Device, SwitchConfig);
+            MainWindow.ApplyTheme();
```
Add empty line before.

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,431 @@
+
+        public readonly Dictionary<string, string> GdkToOpenTKInput = new Dictionary<string, string>()
+        {
+            { "Key_0"      , "Number0"         },
```
```suggestion
            { "Key_0",       "Number0"         },
```

The alignment should be after the comma.

Xpl0itR commented on Ryujinx.HLE/HOS/Horizon.cs:
```diff
@@ -687,7 +702,7 @@ protected virtual void Dispose(bool disposing)
                 // have exited.
                 ThreadCounter.Signal();
-                ThreadCounter.Wait();
+                //ThreadCounter.Wait();
```
Because it is a temporary fix to a bug that i don't know how to fix

Thog replied:
Maybe add TODO here to fix it

Xpl0itR commented:
Forgot to make this comment 2 days ago, Ac_k's requested changes have been changed

AcK77 commented on Ryujinx/Program.cs:
```diff
@@ -1,169 +1,41 @@
-                device.Dispose();
-            }
+            Application gtkApplication = new Application("Ryujinx.Ryujinx", GLib.ApplicationFlags.None);
+            MainWindow mainWindow      = new MainWindow(args, gtkApplication);
```
```suggestion
            MainWindow  mainWindow     = new MainWindow(args, gtkApplication);
```

AcK77 commented on Ryujinx/Ui/AboutWindow.cs:
```diff
@@ -0,0 +1,116 @@
+
+            try
+            {
+                var resolver = CompositeResolver.Create(new[] { StandardResolver.AllowPrivateSnakeCase });
```
maybe an explicit cast for `resolver` is better for consistency?

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,448 @@
+        private static Keyset KeySet;
+        private static SystemState.TitleLanguage DesiredTitleLanguage;
+
+        private const double SecondsPerMinute = 60.0;
```
nit: Align assignments

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,448 @@
+
+                                if (applicationIcon == null)
+                                {
+                                    if (Path.GetExtension(applicationPath) == ".xci")
```
Since you use this same condition third times (L185, L203 & L221). Why don't use a little helper method instead of use the same code?

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,448 @@
+                                    titleId = controlData.SaveDataOwnerId.ToString("x16");
+                                }
+
+                                if (string.IsNullOrWhiteSpace(titleId)) {
```
Nit: bracket on a new line

AcK77 commented on Ryujinx/Ui/ApplicationLibrary.cs:
```diff
@@ -0,0 +1,448 @@
+        {
+            try
+            {
+                string[] playedData = new string[2];
```
Nit: Alignments of vars names and assignments

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,616 @@
+        {
+            MessageDialog errorDialog = new MessageDialog(null, DialogFlags.Modal, MessageType.Error, ButtonsType.Ok, errorMessage);
+            errorDialog.SetSizeRequest(100, 20);
+            errorDialog.Title = "Ryujinx - Error";
```
Align assignments of those 3 next lines

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,616 @@
+                        details = $"Playing {_device.System.TitleName}";
+                    }
+
+                    DiscordPresence.Details = details;
```
Nit: align assignments

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,616 @@
+                        Directory.CreateDirectory(savePath);
+                        using (FileStream stream = File.OpenWrite(System.IO.Path.Combine(savePath, "TimePlayed.dat"))) { stream.Write(Encoding.ASCII.GetBytes("0")); }
+                    }
+                    if (File.Exists(System.IO.Path.Combine(savePath, "LastPlayed.dat")) == false)
```
Add a new line before

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,616 @@
+                        Directory.CreateDirectory(savePath);
+                        using (FileStream stream = File.OpenWrite(System.IO.Path.Combine(savePath, "LastPlayed.dat"))) { stream.Write(Encoding.ASCII.GetBytes("Never")); }
+                    }
+                    using (FileStream stream = File.OpenWrite(System.IO.Path.Combine(savePath, "LastPlayed.dat")))
```
Add a new line before

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,616 @@
+                    }
+                    using (FileStream stream = File.OpenWrite(System.IO.Path.Combine(savePath, "LastPlayed.dat")))
+                    {
+                        using (StreamWriter sr = new StreamWriter(stream))
```
`sr` > `writer`

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,616 @@
+
+        private void Load_Application_File(object o, EventArgs args)
+        {
+            FileChooserDialog fc = new FileChooserDialog("Choose the file to open", this, FileChooserAction.Open, "Cancel", ResponseType.Cancel, "Open", ResponseType.Accept);
```
`fc` > `fileChooser`

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,616 @@
+
+        private void Load_Application_Folder(object o, EventArgs args)
+        {
+            FileChooserDialog fc = new FileChooserDialog("Choose the folder to open", this, FileChooserAction.SelectFolder, "Cancel", ResponseType.Cancel, "Open", ResponseType.Accept);
```
`fc` > `fileChooser`

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,616 @@
+
+        private void FullScreen_Toggled(object o, EventArgs args)
+        {
+            if (_fullScreen.Active == true)
```
`== true` isn't needed here

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,616 @@
+            SettingsWin.Show();
+        }
+
+        private void Nfc_Pressed(object o, EventArgs args)
```
Since we don't plan do emulate NFC related things soon. I think you can remove this for now. We will see that later.

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,616 @@
+            AboutWin.Show();
+        }
+
+        private void Icon_Toggled(object o, EventArgs args)
```
You don't resolve any `Toggled` method here... This condition is not really needed... You can assign the bool directly.

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,424 @@
+
+        public readonly Dictionary<string, string> GdkToOpenTKInput = new Dictionary<string, string>()
+        {
+            { "Key_0",       "Number0"         },
```
```suggestion
            { "Key_0",       "Number0"        },
```
There is an extra space on all thoses following lines.

Xpl0itR commented:
requested changes changed

`When a tabs is active, the text color is wrong and the text is unreadable. Maybe it's because the color is set by the system? But since the background color is set by the theme, the text color should be set by the theme too.`
I will have to spend some time with the css file to change the color

`Load Application from Folder should be a little more explicit or something. Because a new user can't understand it's mean an "unpacked game".`
A new user wouldn't really be loading unpacked games imho

`In the About window, there are a bunch of web links, but the cursor should be changed by a hand (or something similar) when you hover them, because if not, you can't know there is a link under your cursor.`
I will took into this and possibly implement in the next PR

Xpl0itR commented:
Fixed the issue with default theme, Everything should be resolved now

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,595 @@
+                    if (File.Exists(System.IO.Path.Combine(savePath, "TimePlayed.dat")) == false)
+                    {
+                        Directory.CreateDirectory(savePath);
+                        using (FileStream stream = File.OpenWrite(System.IO.Path.Combine(savePath, "TimePlayed.dat"))) { stream.Write(Encoding.ASCII.GetBytes("0")); }
```
Nit: Expand brackets

AcK77 commented on Ryujinx/Ui/MainWindow.cs:
```diff
@@ -0,0 +1,595 @@
+                    if (File.Exists(System.IO.Path.Combine(savePath, "LastPlayed.dat")) == false)
+                    {
+                        Directory.CreateDirectory(savePath);
+                        using (FileStream stream = File.OpenWrite(System.IO.Path.Combine(savePath, "LastPlayed.dat"))) { stream.Write(Encoding.ASCII.GetBytes("Never")); }
```
Nit: Expand brackets

AcK77 commented on Ryujinx/Ui/SwitchSettings.cs:
```diff
@@ -0,0 +1,424 @@
+
+            _gameDirsBox.AppendColumn("", new CellRendererText(), "text", 0);
+            _gameDirsBoxStore   = new ListStore(typeof(string));
+            _gameDirsBox.Model  = _gameDirsBoxStore;
```
```suggestion
            _gameDirsBox.Model = _gameDirsBoxStore;
```
And update the line before too

Xpl0itR commented:
fixed a bug and completed requested changes

Xpl0itR commented:
`Yeah but since the text isn’t really clear, I can thinking it’s for open his Games folder or something like that... As I said, we need to find better word to avoid any confusion.`
OK I'll change that

`What about the fullscreen toggle? `
This will resolve itself once games get rendered to the GLArea in the GUI rather than the new window

`there is some issues with tooltips text too`
I haven't noticed this, I will look into it

Xpl0itR commented:
fixed
